### PR TITLE
Preventing antagonists from spawning on the CC

### PR DESCRIPTION
**Pull request by VigersRay**
## About the PR

Fixes the spawning of spiders, slimes and rats on the CC, as well as the prevention of endowing the mind with automata in the same place.

**Media**

- [ ] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl: VigersRay
- fix: Ghost role antagonists no longer spawn on the CentCom station.

### CODE REVIEW

**Comment by crazybrain23** [comment]
Try to not use uncommon acronyms in changelogs, it makes it hard to understand.

**Comment by VigersRay** [comment]
> Try to not use uncommon acronyms in changelogs, it makes it hard to understand.

I'm bad at English, can you write the best way to write it?

**Comment by crazybrain23** [comment]
> > Try to not use uncommon acronyms in changelogs, it makes it hard to understand.
> 
> I'm bad at English, can you write the best way to write it?

Provided I understood what this PR is doing:
"Fix: Ghost role antagonists no longer spawn on the CentCom station."
Or
"Fix: Ghost role antags no longer spawn at CentCom."

File: Content.Server/StationEvents/Events/MouseMigrationRule.cs [hunk 1]
```diff
@@ -1,19 +1,39 @@
...
     {
         base.Started(uid, component, gameRule, args);
 
+        var targetStation = _stationSystem.GetStations().FirstOrNull();
+
+        if (!TryComp(targetStation, out StationDataComponent? data))
+        {
+            Logger.Info("TargetStation not have StationDataComponent");
+            return;
+        }
```
Comment (metalgearsloth) [inline Content.Server/StationEvents/Events/MouseMigrationRule.cs:25]: 1. This should be targeting any station, not the first.
2. This is just a Resolve.

File: Content.Server/StationEvents/Events/MouseMigrationRule.cs [hunk 2]
```diff
@@ -1,19 +1,39 @@
...
         var modifier = GetSeverityModifier();
 
         var spawnLocations = EntityManager.EntityQuery<VentCritterSpawnLocationComponent, TransformComponent>().ToList();
+
+        var grids = data.Grids.ToHashSet();
+        spawnLocations.RemoveAll(
+            backupSpawnLoc =>
+                backupSpawnLoc.Item2.GridUid.HasValue &&
+                !grids.Contains(backupSpawnLoc.Item2.GridUid.Value));
+
```
Comment (metalgearsloth) [inline Content.Server/StationEvents/Events/MouseMigrationRule.cs:36]: This LINQ is significantly slower, please just do it manually.

File: Content.Server/StationEvents/Events/RandomSentienceRule.cs [hunk 3]
```diff
@@ -2,19 +2,38 @@
...
         var mod = GetSeverityModifier();
-        var targetList = EntityQuery<SentienceTargetComponent>().ToList();
+        var targetList = EntityQuery<SentienceTargetComponent, TransformComponent>().ToList();
+
+
+        var grids = data.Grids.ToHashSet();
+        targetList.RemoveAll(
+            backupSpawnLoc =>
+                backupSpawnLoc.Item2.GridUid.HasValue && !grids.Contains(backupSpawnLoc.Item2.GridUid.Value));
+
```
Comment (metalgearsloth) [inline Content.Server/StationEvents/Events/RandomSentienceRule.cs:36]: As above, also do not just copypaste the code onto 3 systems as it triples the work in maintaining it, add a method somewhere.

File: Content.Server/StationEvents/Events/VentCrittersRule.cs [hunk 4]
```diff
@@ -11,16 +14,31 @@ public sealed class VentCrittersRule : StationEventSystem<VentCrittersRuleCompon
...
         // then just iterate until we get it
         // This will be on average twice as fast.
-        var spawnLocations = EntityManager.EntityQuery<VentCritterSpawnLocationComponent>().ToList();
+        var spawnLocations = EntityManager.EntityQuery<VentCritterSpawnLocationComponent, TransformComponent>().ToList();
+
+        var grids = data.Grids.ToHashSet();
+        spawnLocations.RemoveAll(
+            backupSpawnLoc =>
+                backupSpawnLoc.Item2.GridUid.HasValue && !grids.Contains(backupSpawnLoc.Item2.GridUid.Value));
+
```
Comment (metalgearsloth) [inline Content.Server/StationEvents/Events/VentCrittersRule.cs:41]: .

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Server/StationEvents/Events/MouseMigrationRule.cs (hunk 1 above)
Comment (VigersRay) [reply Content.Server/StationEvents/Events/MouseMigrationRule.cs:25]: I don't really understand what the second paragraph means.

Comment (metalgearsloth) [reply Content.Server/StationEvents/Events/MouseMigrationRule.cs]: Use Resolve and pass in the datacomponent if you want to check if it exists and log for it.

**Review by AJCM-git — CHANGES_REQUESTED** [review]
This is a bad way to accomplish this, the main StationSystem should never have to know what others events do. One good way to do this would be adding a new List<string>(custom type list id serializer for `GameMapPrototype`) datafield called `BlacklistedStations` or something to `StationEventComponent`, then in `EventManagerSystem.CanRun` you compare the id of the map the event entity is on with the new field, and finally you set the new datafield to "Centcomm" in the critters and others prototypes. The only problem would be getting the current map id since CanRun doesnt have the event entity uid as a parameter, but you can figure out a way (either with a query, changing the return value/out of some method inside CanRun or making it take an uid and fixing its usages or something else)

**Comment by VigersRay** [comment]
https://github.com/space-wizards/space-station-14/pull/17833

The spawning issue has been resolved in the already merged pull request, so I'm closing this.

**Pull request closed without merging**